### ISSUE TITLE
withMonthHeaderItemModelProvider working?

### ISSUE THREAD

**Issue by peterboni** [airbnb/HorizonCalendar#73]
Am I doing something wrong?

Minor change to ```SingleDaySelectionDemoViewController``` as follows:

```
...
            .withHorizontalDayMargin(8)
            
            .withMonthHeaderItemModelProvider { [weak self] month in
                let textColor: UIColor
                if #available(iOS 13.0, *) {
                    textColor = .label
                } else {
                    textColor = .black
                }

                let monthAccessibilityText: String?
                monthAccessibilityText = nil
                
                return CalendarItemModel<DayView>(
                    invariantViewProperties: .init(textColor: textColor, isSelectedStyle: false),
                    viewModel: .init(dayText: "Month Header Here", dayAccessibilityText: monthAccessibilityText))
            }
            
            .withDayItemModelProvider { [weak self] day in
...
```

And result - no custom header ...

<img width="383" alt="Screen Shot 2020-11-18 at 6 12 20 pm" src="https://user-images.githubusercontent.com/590995/99516900-afd3f100-29c9-11eb-89fc-714a60ff270a.png">

**Comment by saudeon** [comment]
I'm not sure if you are able to provide the same `CalendarItemModel` for both the day and month.

I do not have this issue with separate models. I created a `MonthLabel` view and it works as expected. I would try that and report back if it is still an issue!

**Comment by peterboni** [comment]
Ok so modified ```HorizonCalendarExample``` slightly more as you suggested @saudeon ...

```
...
            .withHorizontalDayMargin(8)
            
            .withMonthHeaderItemModelProvider { [weak self] month in
                let textColor: UIColor
                if #available(iOS 13.0, *) {
                    textColor = .label
                } else {
                    textColor = .black
                }

                let monthAccessibilityText: String?
                monthAccessibilityText = nil
                
                return CalendarItemModel<MonthHeaderView>(
                    invariantViewProperties: .init(textColor: textColor),
                    viewModel: .init(monthHeaderText: "Month Header Here", monthHeaderAccessibilityText: monthAccessibilityText))
            }
            
            .withDayItemModelProvider { [weak self] day in
...
```

... and then just copied ```DayView``` ...

```
import HorizonCalendar
import UIKit

// MARK: - MonthHeaderView

final class MonthHeaderView: UIView {

  // MARK: Lifecycle

  init(invariantViewProperties: InvariantViewProperties) {
    monthHeaderLabel = UILabel()
    monthHeaderLabel.font = invariantViewProperties.font
    monthHeaderLabel.textAlignment = invariantViewProperties.textAlignment
    monthHeaderLabel.textColor = invariantViewProperties.textColor

    super.init(frame: .zero)

    addSubview(monthHeaderLabel)
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  // MARK: Internal

  var monthHeaderText: String {
    get { monthHeaderLabel.text ?? "" }
    set { monthHeaderLabel.text = newValue }
  }

  var monthHeaderAccessibilityText: String?

  override func layoutSubviews() {
    super.layoutSubviews()
    monthHeaderLabel.frame = bounds
    layer.cornerRadius = min(bounds.width, bounds.height) / 2
  }

  // MARK: Private

  private let monthHeaderLabel: UILabel
}

// MARK: UIAccessibility

extension MonthHeaderView {

  override var isAccessibilityElement: Bool {
    get { true }
    set { }
  }

  override var accessibilityLabel: String? {
    get { monthHeaderAccessibilityText ?? monthHeaderText }
    set { }
  }

}

// MARK: CalendarItemViewRepresentable

extension MonthHeaderView: CalendarItemViewRepresentable {

  struct InvariantViewProperties: Hashable {
    var font = UIFont.systemFont(ofSize: 18)
    var textAlignment = NSTextAlignment.center
    var textColor: UIColor
  }

  struct ViewModel: Equatable {
    let monthHeaderText: String
    let monthHeaderAccessibilityText: String?
  }

  static func makeView(
    withInvariantViewProperties invariantViewProperties: InvariantViewProperties)
    -> MonthHeaderView
  {
    MonthHeaderView(invariantViewProperties: invariantViewProperties)
  }

  static func setViewModel(_ viewModel: ViewModel, on view: MonthHeaderView) {
    view.monthHeaderText = viewModel.monthHeaderText
    view.monthHeaderAccessibilityText = viewModel.monthHeaderAccessibilityText
  }

}
```

... same result ...

<img width="393" alt="image" src="https://user-images.githubusercontent.com/590995/99729637-2d951b00-2af6-11eb-94bd-76d2678554fe.png">

**Comment by saudeon** [comment]
Interesting...

My implementation within my application is quite custom, so I'll see if I can reproduce with the example project in here. Will get back to you ASAP.

**Comment by saudeon** [comment]
Okay, so I attempted to do it with a modified `DayLabel` and received a similar issue.

I created the following for a simple custom Month label.

```
struct MonthLabel: CalendarItemViewRepresentable {

  /// Properties that are set once when we initialize the view.
  struct InvariantViewProperties: Hashable {
    let font: UIFont
    var textColor: UIColor
    var backgroundColor: UIColor
  }

  /// Properties that will vary depending on the particular date being displayed.
  struct ViewModel: Equatable {
    let month: Month
  }

  static func makeView(withInvariantViewProperties invariantViewProperties: InvariantViewProperties) -> UILabel {
    let label = UILabel()

    label.backgroundColor = invariantViewProperties.backgroundColor
    label.font = invariantViewProperties.font
    label.textColor = invariantViewProperties.textColor
    
    label.textAlignment = .center
    
    return label
  }

  static func setViewModel(_ viewModel: ViewModel, on view: UILabel) {
    view.text = "\(viewModel.month)"
  }

}
```

And then I modified the `withMonthHeaderItemModelProvider` to look like the following.
```
.withMonthHeaderItemModelProvider { [weak self] month in
        let textColor: UIColor
        if #available(iOS 13.0, *) {
          textColor = .label
        } else {
          textColor = .black
        }

        return CalendarItemModel<MonthLabel>(
            invariantViewProperties: .init(font: UIFont.systemFont(ofSize: 12.0),
                                           textColor: textColor,
                                           backgroundColor: .clear),
          viewModel: .init(month: month))
     }
```

And that worked as expected. 
I'm sure someone else can chime in and make sense of why your original attempt did not work!

**Comment by peterboni** [comment]
@saudeon confirmed your work-around works. Thanks!
Looks like a bug trying to do it the recommended way then.

**Comment by saudeon** [comment]
Glad that I could help, hopefully we can gain some insight into what caused your initial implementation to not work!

**Comment by bryankeller** [comment]
@peterboni / @saudeon the issue here is that month headers are self-sized (their height is determined by their intrinsic height / the constraints of any subviews of the view). Day views are not self-sized (their width and height are determined by the available size of the calendar and any spacing values provided via things like `withHorizontalDayMargin`). 

If you had modified the `MonthHeaderView` example to constrain the `monthHeaderLabel` subview to the edges of the superview, and deleted the `layoutSubview` override, it would have worked correctly.

This is definitely a failure of the API. This behavior is not clear without digging into the internal source code. It probably makes sense for me to add in an explicit sizing approach for month headers so that this behavior is explicit / not confusing.

**Comment by saudeon** [comment]
Ah, this makes more sense now! Thanks for the clarification, I assumed it was due to the `layoutSubview` override but wasn't 100% sure why!

**Comment by peterboni** [comment]
Thanks @bryankeller.

Awesome calendar control BTW. Thank you for releasing it!

**Comment by druvinskiy** [comment]
> .withMonthHeaderItemModelProvider { [weak self] month in
>         let textColor: UIColor
>         if #available(iOS 13.0, *) {
>           textColor = .label
>         } else {
>           textColor = .black
>         }
> 
>         return CalendarItemModel<MonthLabel>(
>             invariantViewProperties: .init(font: UIFont.systemFont(ofSize: 12.0),
>                                            textColor: textColor,
>                                            backgroundColor: .clear),
>           viewModel: .init(month: month))
>      }

Hello. I tried using your method in m
![Simulator Screen Shot - iPhone 12 Pro - 2020-12-21 at 13 48 56](https://user-images.githubusercontent.com/19519587/102816280-5f1e3200-4393-11eb-8d1c-e781d5c29b5d.png)
y app. However, for some reason my month is in the "2020-01" rather than in the "January 2020" format. Is there something I'm missing to get it to show up in the correct format?
![Simulator Screen Shot - iPhone 12 Pro - 2020-12-21 at 13 48 56](https://user-images.githubusercontent.com/19519587/102816255-57f72400-4393-11eb-99bd-210a83ccfb0d.png)

**Comment by bryankeller** [comment]
@druvinskiy can you please make a new issue for this?
